syntax = "proto3";

package demoshop;

// ===============
// Common messages
// ===============
message Empty {}

// Represents an amount of money with its currency type.
// ie: $1.75 USD: current_code = USD, units = -1, nanos = -750,000,000
message Money {
  // 3-letter currency code defined in ISO 4217
  string currency_code = 1;
  // The whole units of the amount.
  int64 units = 2;
  // Between -999,999,999 and +999,999,999.
  int32 nanos = 3;
}

// ===============
// Product Catalog
// ===============
service ProductCatalogService {
  rpc ListProducts(Empty) returns (ListProductsResponse) {}
  rpc GetProduct(GetProductRequest) returns (Product) {}
  rpc SearchProducts(SearchProductsRequest) returns (SearchProductsResponse) {}
}

message Product {
  string id = 1;
  string name = 2;
  string description = 3;
  string picture = 4;
  Money price_usd = 5;
  repeated string categories = 6;
}

message ListProductsResponse {
  repeated Product products = 1;
}

message GetProductRequest {
  string id = 1;
}

message SearchProductsRequest {
  string query = 1;
}

message SearchProductsResponse {
  repeated Product results = 1;
}

// ======================
// Recommendation Service
// ======================
service RecommendationService {
  rpc ListRecommendations(ListRecommendationsRequest) returns (ListRecommendationsResponse) {}
}

message ListRecommendationsRequest {
  string user_id = 1;
  repeated string product_ids = 2;
}

message ListRecommendationsResponse {
  repeated string product_ids = 1;
}
